File: str-comb/scomb.cpp
//////////////////////////////
// scomb.cpp
#include<cstdio>
#include<string>
using namespace std;
int newGap(int gap){
    gap /= 1.3;
    if(gap == 9 || gap == 10) return 11;
    if(gap < 1) return 1;
    return gap;
}
void combSort(string a[], int len){
    int gap = len;
    bool swapped;
    do {
        swapped = false;
        gap = newGap(gap);
        for(int i=0; i < len-gap; ++i) {
            if(a[i] > a[i+gap]) {
                swapped = true;
                a[i].swap(a[i+gap]);
            }
        }
    } while(gap > 1 || swapped);
}
const int N = 10000000;
int main() {
    string* arr = new string[N];
    for(int z=0;z<N;++z) arr[z] = to_string(N-z);    
    combSort(arr,N);
    for(int z=1;z<N;++z) if(arr[z]<arr[z-1]) printf("!");
    delete[] arr;
}

File: str-comb/scomb.jl
##############################
# scomb.jl
function newGap(gap)
    gap = floor(Int, gap / 1.3)
    if gap == 9 || gap == 10; return 11 end
    if gap < 1; return 1 end
    gap
end
function combSort!(a, len)
    gap = len
    while true
        swapped = false
        gap = newGap(gap)
        for i in 1:(len-gap)
            j = i + gap
            if a[i] > a[j]
                swapped = true
                a[i], a[j] = a[j], a[i]
            end
        end
        if gap <= 1 && !swapped; break end
    end
end
const N = 10_000_000
function main()
    arr = [ string(N-i+1) for i in 1:N ]
    combSort!(arr, N)
    for i in 2:N
        if arr[i] < arr[i-1]
            print("!")
        end
    end
end

main()
